crates/csm-worker/src/processor.rs

```diff
@@ -0,0 +1,163 @@
+    operation::{ClientUpdateOutput, SyncAction},
+};
+use tracing::*;
+
+use crate::{state::CsmWorkerState, sync_actions::apply_action};
+
+/// Process a single ASM log entry, extracting and handling checkpoint updates.
+pub(crate) fn process_checkpoint_log(
+    state: &mut CsmWorkerState,
+    log: &AsmLogEntry,
+    asm_block: &L1BlockCommitment,
+) -> anyhow::Result<()> {
+    // Check if this is a checkpoint update log
+    if log.ty() != Some(CHECKPOINT_UPDATE_LOG_TYPE) {
+        trace!("Skipping non-checkpoint log");
+        return Ok(());
+    }
+
+    // Deserialize the checkpoint update using the AsmLog trait
+    let checkpoint_update: CheckpointUpdate = log
+        .try_into_log()
+        .map_err(|e| anyhow::anyhow!("Failed to deserialize CheckpointUpdate: {}", e))?;
+
+    let epoch = checkpoint_update.batch_info.epoch();
+
+    info!(
+        %epoch,
+        %asm_block,
+        checkpoint_txid = ?checkpoint_update.checkpoint_txid,
+        "Processing checkpoint update from ASM log"
+    );
+
+    // Create L1 checkpoint reference from the log data
+    let l1_reference = CheckpointL1Ref::new(
+        *asm_block,
+        checkpoint_update.checkpoint_txid.inner_raw(),
+        checkpoint_update.checkpoint_txid.inner_raw(), // TODO: get wtxid if available
+    );
+
+    // Create L1Checkpoint for client state
+    let l1_checkpoint = L1Checkpoint::new(
+        checkpoint_update.batch_info.clone(),
+        BatchTransition {
+            epoch,
+            chainstate_transition: checkpoint_update.chainstate_transition,
+        },
+        l1_reference.clone(),
+    );
+
+    // Update the client state with this checkpoint
+    update_client_state_with_checkpoint(state, l1_checkpoint, epoch)?;
+
+    // Create sync action to update checkpoint entry in database
+    let sync_action = SyncAction::UpdateCheckpointInclusion {
+        checkpoint: create_checkpoint_from_update(&checkpoint_update),
+        l1_reference,
+    };
+
+    // Apply the sync action
+    apply_action(sync_action, &state.storage)?;
+
+    // Track the last processed epoch
+    state.last_processed_epoch = Some(epoch);
+
+    Ok(())
+}
+
+/// Update client state with a new checkpoint.
+fn update_client_state_with_checkpoint(
+    state: &mut CsmWorkerState,
+    new_checkpoint: L1Checkpoint,
+    epoch: u64,
+) -> anyhow::Result<()> {
+    // Get the current client state
+    let cur_state = state.cur_state.as_ref();
+
+    // Determine if this checkpoint should be the last finalized or just recent
+    let (last_finalized, recent) = match cur_state.get_last_checkpoint() {
+        Some(existing) => {
+            // If the new checkpoint is for a later epoch, it becomes recent
+            if epoch > existing.batch_info.epoch() {
+                (Some(existing.clone()), Some(new_checkpoint))
+            } else {
+                // Otherwise keep existing
+                (Some(existing.clone()), None)
```

Comment:
Why do we move the existing checkpoint to be final? Shouldn't this depend on the number of block confirmations?